### PR TITLE
Performance improvements 1

### PR DESCRIPTION
**Pull request by fabianmurariu**
treat the edges of a vertex as views instead of copying lists
remove synchronised from accumulators and replaced them with atomic operations
various other changes ..

### CODE REVIEW

File: core/src/main/scala/com/raphtory/api/analysis/graphstate/Accumulator.scala [hunk 1]
```diff
@@ -8,7 +8,7 @@ package com.raphtory.api.analysis.graphstate
   *
   * @see [[GraphState]]
   */
-trait Accumulator[-S, T] {
+trait Accumulator[@specialized(Int) -S, T] {
```
Comment (Alnaimi-) [inline core/src/main/scala/com/raphtory/api/analysis/graphstate/Accumulator.scala:11]: Do we want this to extend beyond Int?

So @specialized(Int, Long, Boolean, ...)

Comment (fabianmurariu) [reply core/src/main/scala/com/raphtory/api/analysis/graphstate/Accumulator.scala:11]: Yes, it's WIP still

Comment (fabianmurariu) [reply core/src/main/scala/com/raphtory/api/analysis/graphstate/Accumulator.scala:11]: It's specialised now

File: core/src/main/scala/com/raphtory/api/analysis/graphstate/AccumulatorImplementation.scala [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+    val noBins: Int,
+    override val minValue: T,
+    override val maxValue: T
+) extends Histogram[T] {
+  val bins: Array[Int] = Array.fill(noBins)(0)
+  var totalCount: Int  = 0
+
+  override def getBins: Array[Int] = bins
+
+  override def cumSum: Array[Int] = bins.scanLeft(0)(_ + _)
```
Comment (Alnaimi-) [inline core/src/main/scala/com/raphtory/api/analysis/graphstate/AccumulatorImplementation.scala:160]: Weird name...

File: core/src/main/scala/com/raphtory/algorithms/generic/dynamic/WeightedRandomWalk.scala [hunk 3]
```diff
@@ -17,8 +17,9 @@ class WeightedRandomWalk[T: Numeric](
     if (neighbours.isEmpty)
       vertex.ID
     else {
-      val weights = vertex.outEdges.map(e => e.weight[T](weight).toDouble)
-      neighbours(rnd.sample(weights))
+      val weights = vertex.outEdges.map(e => e.weight[T](weight).toDouble).toVector
+      val i = rnd.sample(weights)
+      neighbours.zipWithIndex.find{case (_, j) => i == j}.get._1
```
Comment (Alnaimi-) [inline core/src/main/scala/com/raphtory/algorithms/generic/dynamic/WeightedRandomWalk.scala:22]: Should this possibly move down to the `neighbours` implementation.

File: core/src/main/scala/com/raphtory/api/analysis/graphstate/AccumulatorImplementation.scala (hunk 2 above)
Comment (fabianmurariu) [reply core/src/main/scala/com/raphtory/api/analysis/graphstate/AccumulatorImplementation.scala:160]: that's not me :laughing:

Comment (narnolddd) [reply core/src/main/scala/com/raphtory/api/analysis/graphstate/AccumulatorImplementation.scala:160]: 😳

Comment (fabianmurariu) [reply core/src/main/scala/com/raphtory/api/analysis/graphstate/AccumulatorImplementation.scala:160]: We found him! :laughing: 

https://github.com/Raphtory/Raphtory/blob/development/core/src/main/scala/com/raphtory/api/analysis/graphstate/Histogram.scala#L35

File: core/src/main/scala/com/raphtory/api/analysis/visitor/InterlayerEdge.scala [hunk 4]
```diff
@@ -66,6 +66,7 @@ case class InterlayerEdge(
   override def containsState(key: String, includeProperties: Boolean): Boolean = ???
   override def getOrSetState[T](key: String, value: T, includeProperties: Boolean): T = ???
   override def appendToState[T: ClassTag](key: String, value: T): Unit = ???
+  override def clearState(key: String): Unit = ???
```
Comment (miratepuffin) [inline core/src/main/scala/com/raphtory/api/analysis/visitor/InterlayerEdge.scala:69]: This class needs some serious love xD

Comment (narnolddd) [reply core/src/main/scala/com/raphtory/api/analysis/visitor/InterlayerEdge.scala:69]: agree haha, it's more that I think we'd need a discussion on what each of these things (mainly getState) would mean for an interlayer edge

File: core/src/main/scala/com/raphtory/api/analysis/graphstate/GraphStateImplementation.scala [hunk 5]
```diff
@@ -199,6 +31,10 @@ private[raphtory] class GraphStateImplementation(override val nodeCount: Int) ex
     accumulatorState(name) = AccumulatorImplementation[T](initialValue, retainState, numeric.plus)
       .asInstanceOf[AccumulatorImplementation[Any, Any]]
 
+  override def newIntAdder(name: String, initalValue: Int, retainState: Boolean): Unit =
```
[the remark this answers is not in the document]
Comment (fabianmurariu) [reply core/src/main/scala/com/raphtory/api/analysis/graphstate/GraphStateImplementation.scala:34]: it's not a high priority but they do show up when we start doing actual work on larger graphs

File: core/src/main/scala/com/raphtory/api/analysis/visitor/InterlayerEdge.scala (hunk 4 above)
Comment (fabianmurariu) [reply core/src/main/scala/com/raphtory/api/analysis/visitor/InterlayerEdge.scala:69]: I think this is on @narnolddd

File: core/src/main/scala/com/raphtory/internals/components/Component.scala [hunk 6]
```diff
@@ -120,7 +120,7 @@ object Component {
     catch {
       case ex: Exception =>
         log.warn("Failed to disable Java 10 access warning:")
-        ex.printStackTrace()
+//        ex.printStackTrace()
```
[the remark this answers is not in the document]
Comment (fabianmurariu) [reply core/src/main/scala/com/raphtory/internals/components/Component.scala:123]: I've been running Raphtory on Java-17 this thing .. doesn't work anymore

**Review by miratepuffin — APPROVED** [review]

**Pull request merged**